src/Plugin/PasswordConstraint/MinimumCharacters.php

```diff
@@ -0,0 +1,102 @@
+<?php
+
+namespace Drupal\password_enhancements\Plugin\PasswordConstraint;
+
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Plugin\PluginBase;
+
+/**
+ * Defines a common class for minimum character requirement.
+ */
+abstract class MinimumCharacters extends PluginBase {
+
+  /**
+   * The generated error message.
+   *
+   * @var string|null
+   */
+  protected $errorMessage;
+
+  /**
+   * {@inheritdoc}
+   */
+  public function __construct(array $configuration, $plugin_id, $plugin_definition) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+
+    $this->errorMessage = NULL;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getErrorMessage(): ?string {
+    return $this->errorMessage;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getInitialDescription(): string {
+    if ($this->configuration['minimum_characters'] > 1) {
+      return $this->configuration['descriptionPlural'];
+    }
+    else {
+      return $this->configuration['descriptionSingular'];
+    }
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function settingsForm(): array {
+    $form = [];
+
+    $form['minimum_characters'] = [
+      '#type' => 'number',
+      '#title' => $this->t('Minimum characters'),
+      '#required' => TRUE,
+      '#min' => 1,
+      '#default_value' => !empty($this->configuration['minimum_characters']) ? $this->configuration['minimum_characters'] : 1,
```

Comment:
This can be `$this->configuration['minimum_characters'] ?? 1`.

Reply:
Let's just eliminate the possibility that the value can be accidentally 0, so it should be checked with empty().

Btw `$var ?? 'default'` is equivalent to `isset($var) ? $var : 'default'` https://www.php.net/manual/en/migration70.new-features.php#migration70.new-features.null-coalesce-op

Reply:
What about negative values? :) Even if the form does not allow to set it, this should be validated somewhere as well because CIM can import anything.

#5